types/mempool/priority_nonce.go

```diff
@@ -0,0 +1,377 @@
+	if res != 0 {
+		return res
+	}
+
+	// weight is used as a tiebreaker for transactions with the same priority.  weight is calculated in a single
+	// pass in .Select(...) and so will be 0 on .Insert(...)
+	res = huandu.Int64.Compare(keyA.weight, keyB.weight)
+	if res != 0 {
+		return res
+	}
+
+	// Because weight will be 0 on .Insert(...), we must also compare sender and nonce to resolve priority collisions.
+	// If we didn't then transactions with the same priority would overwrite each other in the priority index.
+	res = huandu.String.Compare(keyA.sender, keyB.sender)
+	if res != 0 {
+		return res
+	}
+
+	return huandu.Uint64.Compare(keyA.nonce, keyB.nonce)
+}
+
+type PriorityNonceMempoolOption func(*priorityNonceMempool)
+
+// WithOnRead sets a callback to be called when a tx is read from the mempool.
+func WithOnRead(onRead func(tx sdk.Tx)) PriorityNonceMempoolOption {
+	return func(mp *priorityNonceMempool) {
+		mp.onRead = onRead
+	}
+}
+
+// DefaultPriorityMempool returns a priorityNonceMempool with no options.
+func DefaultPriorityMempool() Mempool {
+	return NewPriorityMempool()
+}
+
+// NewPriorityMempool returns the SDK's default mempool implementation which
+// returns txs in a partial order by 2 dimensions; priority, and sender-nonce.
+func NewPriorityMempool(opts ...PriorityNonceMempoolOption) Mempool {
+	mp := &priorityNonceMempool{
+		priorityIndex:  huandu.New(huandu.LessThanFunc(txMetaLess)),
+		priorityCounts: make(map[int64]int),
+		senderIndices:  make(map[string]*huandu.SkipList),
+		scores:         make(map[txMeta]txMeta),
+	}
+
+	for _, opt := range opts {
+		opt(mp)
+	}
+
+	return mp
+}
+
+// Insert attempts to insert a Tx into the app-side mempool in O(log n) time,
+// returning an error if unsuccessful. Sender and nonce are derived from the
+// transaction's first signature.
+//
+// Transactions are unique by sender and nonce. Inserting a duplicate tx is an
+// O(log n) no-op.
+//
+// Inserting a duplicate tx with a different priority overwrites the existing tx,
+// changing the total order of the mempool.
+func (mp *priorityNonceMempool) Insert(ctx sdk.Context, tx sdk.Tx) error {
+	sigs, err := tx.(signing.SigVerifiableTx).GetSignaturesV2()
+	if err != nil {
+		return err
+	}
+	if len(sigs) == 0 {
+		return fmt.Errorf("tx must have at least one signer")
+	}
+
+	priority := ctx.Priority()
+	sig := sigs[0]
+	sender := sig.PubKey.Address().String()
+	nonce := sig.Sequence
+	key := txMeta{nonce: nonce, priority: priority, sender: sender}
+
+	senderIndex, ok := mp.senderIndices[sender]
+	if !ok {
+		senderIndex = huandu.New(huandu.LessThanFunc(func(a, b any) int {
+			return huandu.Uint64.Compare(b.(txMeta).nonce, a.(txMeta).nonce)
+		}))
+
+		// initialize sender index if not found
+		mp.senderIndices[sender] = senderIndex
+	}
+
+	mp.priorityCounts[priority]++
+
+	// Since senderIndex is scored by nonce, a changed priority will overwrite the
+	// existing key.
+	key.senderElement = senderIndex.Set(key, tx)
```

Comment:
if we overwrite the tx with the changed priority, should we do `mp.priorityCounts[priority]--`?